qiskit/qasm3/exporter.py

```diff
@@ -196,139 +196,308 @@ def dump(self, circuit, stream):
+            else:
+                self.user_gate_idents[id(source)] = ident
+        return ident
 
-# A _Scope is the structure used in the builder to store the contexts and re-mappings of bits from
-# the top-level scope where the bits were actually defined.  In the class, 'circuit' is an instance
-# of QuantumCircuit that defines this level, and 'bit_map' is a mapping of 'Bit: Bit', where the
-# keys are bits in the circuit in this scope, and the values are the Bit in the top-level scope in
-# this context that this bit actually represents.  'symbol_map' is a bidirectional mapping of
-# '<Terra object>: Identifier' and 'str: <Terra object>', where the string in the second map is the
-# name of the identifier.  This is a cheap hack around actually implementing a proper symbol table.
-_Scope = collections.namedtuple("_Scope", ("circuit", "bit_map", "symbol_map"))
+    def register_gate(
+        self,
+        name: str,
+        source: Gate,
+        params: Iterable[ast.Identifier],
+        qubits: Iterable[ast.Identifier],
+        body: ast.QuantumBlock,
+    ) -> ast.Identifier:
+        """Register the given gate in the symbol table, using the given components to build up the
+        full AST definition."""
+        name = self.escaped_declarable_name(name, allow_rename=True)
+        ident = ast.Identifier(name)
+        self.gates[name] = GateDefinition(
+            source, ast.QuantumGateDefinition(ident, tuple(params), tuple(qubits), body)
+        )
+        # Add the gate object with a magic lookup keep to the objects dictionary so we can retrieve
+        # it later.  Standard gates are not guaranteed to have stable IDs (they're preferentially
+        # not even created in Python space), but user gates are.
+        if (standard_gate := getattr(source, "_standard_gate")) is not None:
+            self.standard_gate_idents[standard_gate] = ident
+        else:
+            self.user_gate_idents[id(source)] = ident
+        return ident
+
+    def get_gate(self, gate: Gate) -> ast.Identifier | None:
+        """Lookup the identifier for a given `Gate`, if it exists."""
+        source = _gate_definition_source(gate)
+        # `our_defn.source` means a basis gate that we should assume is always valid.
+        if (our_defn := self.gates.get(gate.name)) is not None and (
+            our_defn.source is None or our_defn.source == source
+        ):
+            return ast.Identifier(gate.name)
+        if (standard_gate := getattr(source, "_standard_gate")) is not None:
+            if (our_ident := self.standard_gate_idents.get(standard_gate)) is None:
+                return None
+            return our_ident if self.gates[our_ident.string].source == source else None
+        # No need to check equality if we're looking up by `id`; we must have the same object.
+        return self.user_gate_idents.get(id(source))
+
+
+def _gate_definition_source(gate: Gate) -> Gate:
```

Comment:
I've been wondering why you don't store something like `(GateClass, num_params)`, instead of constructing a gate with fixed, otherwise useless, params. It looks like there are several reasons. One is that we don't have a convenient interface for obtaining information about a gate other than calling methods on instances of the gate. So, you'd have to construct a gate anyway to get it. Another is you don't try to do anything clever with user gates; they are stored as instances, and are their own source, and you hope for the best. So if you treated standard gates differently, your definition source would have a more heterogeneous format. Although this is a sort of vague view of the problem, trying to store the definition source differently seems to cause problems, and not to really solve any.

What you have here is pretty straightforward, given the constraints of current qiskit data models.

Reply:
Hopefully this is a little clearer with the renaming of this method to `_gate_canonical_form`.  It's clearer for typing and usage if everything's always just a `Gate` (or `None`) - no need to do type reflection on lookups, just use the same comparisons as always.